### PR TITLE
Feature/swapaxes

### PR DESCRIPTION
**Pull request by blackbox-tech**
This is for #3891 
I don't know numba well, nor do I fully understand it, so I suspect I may have put this code in the wrong location, feel free to correct or reject this change, no offense will be taken.

### CODE REVIEW

**Comment by esc** [comment]
@blackbox-tech thank you for submitting this, I have added it to the queue for review.

**Comment by stuartarchibald** [comment]
Thanks for the PR. There's a couple of flake8 violations that could do with a fix ahead of review please, just so CI passes. https://dev.azure.com/numba/numba/_build/results?buildId=6653&view=logs&j=9b1c419d-97bd-5aaf-47ca-b83be1ff9bc1&t=d1ed00e6-cc06-521a-e00c-4f5e4a0f55c5&l=85

Many thanks.

**Comment by blackbox-tech** [comment]
Sure will do, give me a few days as I'm snowed under with some other tasks ATM (in the interim, feel free to unqueue this code review if its holding anything up).

**Comment by blackbox-tech** [comment]
@stuartarchibald These Flake8 tests pass now

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the PR, looks like a great start. I've given this patch an initial review and left some comments in line. Thanks again!

**Comment by blackbox-tech** [comment]
Sure I'll follow up next week as I'm on a wilderness canoeing expedition ATM.

**Comment by blackbox-tech** [comment]
@stuartarchibald I've made some changes following your code review.

**Comment by blackbox-tech** [comment]
@stuartarchibald, @esc is this good to go?

**Comment by esc** [comment]
@blackbox-tech thank you for asking about this. We are currently in a release candidate phase and will delay processing this until after the next release candidate is out the door!

**Comment by blackbox-tech** [comment]
Thanks for the update

File: numba/np/arrayobj.py [hunk 1]
```diff
@@ -1542,6 +1543,33 @@ def np_transpose_impl(a, axes=None):
     return np_transpose_impl
 
 
+@overload(np.swapaxes)
+def numpy_swapaxes(a, axis1, axis2):
+    force_literal = lambda a, axis1, axis2: (
+        a, literally(axis1), literally(axis2))
+    if isinstance(axis1, types.Literal):
+        axis1_val = axis1.literal_value
+    else:
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/np/arrayobj.py:1556]: This needs addressing.

File: numba/np/arrayobj.py [hunk 2]
```diff
@@ -1542,6 +1543,33 @@ def np_transpose_impl(a, axes=None):
...
+        axis2 = a.ndim + axis2_val if axis2_val < 0 else axis2_val
+        if axis1 == axis2:
+            return a
+        pos = to_fixed_tuple(np.arange(a.ndim), a.ndim)
+        tmp = pos[axis1]
+        pos = tuple_setitem(pos, axis1, pos[axis2])
+        pos = tuple_setitem(pos, axis2, tmp)
+        return a.transpose(pos)
+
+    return impl
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/np/arrayobj.py]: This needs addressing.

File: numba/tests/test_array_manipulation.py [hunk 3]
```diff
@@ -584,6 +588,38 @@ def check(pyfunc):
...
+        got = cfunc(a, axis1, axis2)
+        np.testing.assert_equal(expected, got)
+
+        # check non-int axis raise an error
+        axis1 = 0
+        axis2 = 1.5
+        pyfunc = numpy_swapaxes_array
+        cr = compile_isolated(pyfunc, (typeof(a), typeof(axis1), typeof(axis2)), flags=flags)
+        cfunc = cr.entry_point
+        with self.assertRaises(TypeError):
```
Comment (stuartarchibald) [inline numba/tests/test_array_manipulation.py:620]: Check out the pattern used in Numba for this, there's an example  few lines down in `test_squeeze_array_npm`. Numba does `with self.assertRaises(<exception>) as e:` and then later checks that there's the expected error message in `e`.

File: numba/np/arrayobj.py [hunk 4]
```diff
@@ -1542,6 +1543,40 @@ def np_transpose_impl(a, axes=None):
...
+        axis2_val = axis2.literal_value
+    elif not isinstance(axis2, types.Integer):
+        raise errors.TypingError('axis2 must be an integer')
+    else:
+        return force_literal
+
+    def impl(a, axis1, axis2):
+        axis1 = a.ndim + axis1_val if axis1_val < 0 else axis1_val
+        axis2 = a.ndim + axis2_val if axis2_val < 0 else axis2_val
+        if axis1 == axis2:
```
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1569]: What if an axis is out of bounds?

File: numba/tests/test_array_manipulation.py [hunk 5]
```diff
@@ -584,6 +588,38 @@ def check(pyfunc):
     def test_transpose_array_npm(self):
         self.test_transpose_array(flags=no_pyobj_flags)
 
+    def test_swapaxes_array(self, flags=enable_pyobj_flags):
```
Comment (stuartarchibald) [inline numba/tests/test_array_manipulation.py:591]: It'd be good to include a case which is the no-op, i.e. axes are equal, and also some different shaped arrays. Also to check errors for out of bounds axis on both axis args.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the fixes so far, they look great. There's a couple of things outstanding from the initial review and a couple of further comments to address regarding error handling and testing. Thanks again!

**Pull request closed without merging**